developer_guides/virtualisation/virtualisation.rst

```diff
@@ -0,0 +1,255 @@
+topology should have no relation to the host hardware. The guest audio topology
+should be decided upon by the host system administrator. Therefore we let the
+SOF instance, running on the guest, obtain topology from the host. Such a
+topology should only contain "software components," like PCMs, buffers, software
+volume controls. Those components thus represent partial pipelines, that then
+will be linked to mixer and demux components of the host SOF topology for
+playback and capture respectively.
+
+With this design the topology is represented in the following way:
+
+DSP: has a full topology, including host and guests' components. No distinction
+is made between the host and any guests.
+
+host: the SOF driver core, running on the host, and the ASoC subsystem only see
+the host topology. The vhost driver manages connection points between the host
+and the guests' topology fragments.
+
+guest: only "sees" the guest parts.
+
+As mentioned above, SOF IPC messages from guests are forwarded exactly 1-to-1 to
+the DSP. E.g., if a guest sends an IPC to adjust audio volume, it will be
+forwarded to the DSP. That IPC contains the volume component ID as seen by the
+respective guest. Therefore it should be exactly the same ID, as the one used by
+the firmware.
+
+3. Implementation
+*****************
+
+3.1. Initialisation
+-------------------
+
+The SOF driver probing on the host remains unmodified, except additionally the
+vhost driver is initialised, which registers a /dev/vhost-dsp misc device.
+
+When QEMU starts a guest instance, it detects the above misc device and
+creates a PCI device for the guest with dedicated vendor and device IDs.
+
+The SOF VirtIO driver on the guest registers support for that PCI device and
+its .probe() method is called.
+
+The SOF driver on the guest performs partial initialisation, it skips any
+steps, involving communicating with the actual DSP, e.g. sending the firmware,
+booting it, tracing initialisation.
+
+The VirtIO driver on the guest uses a new SOF_IPC_TPLG_VFE_GET VirtIO-specific
+IPC message to request the topology from the host. The host reads the respective
+file from its filesystem, and sends it in multiple chunks to the guest in reply
+to that request.
+
+TODO: define a reliable secure topology selection protocol. Currently the guest
+simply sends a configured ID to the host, upon which the host opens the topology
+file with the same name. Currently this isn't a problem, since so far only a
+single guest is supported. So, the host can verify, that the ID, that the guest
+has sent does indeed match with the guest topology file name. However, when
+support for multiple guests is added, it must be ascertained, that guests only
+receive topology files, that they are allowed to get. At the moment there
+doesn't seem to be a way on the host to uniquely and securely identify guests.
+If there is indeed no such way, our only possibility seems to be using
+cryptography for that.
```

Comment:
My expectation is that host FS will contain topology files for guest0, guest1, etc. The guest should identify itself somehow to the host ?

Reply:
That's exactly the point - I don't think there's a standard way to securely do that. We are trying to protect against "fraud," right? I.e. someone installing a rogue VM, which would claim to be "guest0," where in fact it is not. So, as far as I understand the only way to do that is using cryptography. E.g. the host would send a request "identify yourself" along with a pseudo random number as a challenge, the guest could return a signed response "I am guest 0, your challenge was X." But I'm not a cryptography expert and I don't know how much security we really need for this.

Reply:
Lets leave this open atm, guest ID will be trusted so no need for any crypto challenge. Leave room in IPC for guest ID in any init flow.

Reply:
well, it's a major open. If there is no way to identify a guest, then this solution needs to be restricted to a single guest.
Can't we check with virtualization experts what is needed here?

Reply:
I asked on the official QEMU IRC, got no replies. That's certainly not a 100% reliable indication, but usually they're responsive, and they did try to help with this one too initially. But yes, we should ask on the mailing list.

Reply:
No replies on the mailing list either. @lgirdwood what exactly do you mean by leaving room in *any* init flow? We already pass a guest ID / topology file name with "request topology" - is that enough or you mean something different?

Reply:
@lyakh ask Eddie

Reply:
@lgirdwood ok, will do. But I'm wondering: if the topology simply resided in the guest's filesystem, we wouldn't need any authentication? Could you remind me again why this wasn't acceptable, please?

Reply:
guests are not trusted or safety critical certified.

Reply:
ok, yes, my mistake, letting a guest send its own topology would be in no way better than letting it choose which topology it should receive, in fact even worse.

Reply:
@lgirdwood I've got some replies from Eddie and other guys, he involved in the conversation. Basically, I think, what they are proposing, is make per-guest (audio) configuration decision based on guest's instance itself, i.e. eventually on guest's VM disk image. This of course asks for some host user-space support. On the host you would have a configuration file like
```
image: /path/to/vm/images/vm0
audio-topology: /path/to/vm/audio/topology/vm0
```
then SOF support code in QEMU will have to know to send this information to the SOF vhost driver during initialisation. Then the guest SOF virtio driver will just ask for its topology and the SOF vhost driver will already know which one belongs to this VM. Would this be acceptable?

Reply:
Yep, good for me if Eddie approves.

Reply:
> then SOF support code in QEMU will have to know to send this information to the SOF vhost driver during initialisation. 

What is 'this information'? and how would it be sent? and how would the host do a match?

Reply:
@plbossart "QEMU support code" means the SOF patch for QEMU. It opens the /dev/vhost-dsp device, sends ioctl()s to it, configures virtuql queues etc.
"This information" means, once qemu starts with the disk image vm0, that code will check the table in /etc/qemu/driver-conf.d/sof.conf (the path is invented) and send an `ioctl(GUEST_ID, &(struct guest_id){.id = "vm0"});` to the SOF vhost driver.